chameleon/admin.py

```diff
@@ -1,52 +1,156 @@
         keycloak_client = KeycloakClient()
         keycloak_user = keycloak_client.get_user_by_username(obj.requestor.username)
 
         if not keycloak_user:
-            return None
-
-        full_name = '{} {}'.format(keycloak_user['lastName'], keycloak_user['firstName'])
-        email = keycloak_user['email']
-        contents = [f'Name: {full_name}', f'Email: {email}']
-        for key, val in keycloak_user['attributes'].items():
-            if key not in ['joinDate']:
-                #convert camelcase to separate out words
+            return "No Keycloak User Found"
+
+        full_name = "{} {}".format(
+            keycloak_user["lastName"], keycloak_user["firstName"]
+        )
+        email = keycloak_user["email"]
+        contents = [f"Name: {full_name}", f"Email: {email}"]
+
+        if not contents:
+            return "No Keycloak Metadata Found"
+
+        for key, val in keycloak_user["attributes"].items():
+            if key not in ["joinDate"]:
+                # convert camelcase to separate out words
                 key = re.sub("([A-Z])", " \\1", key).strip().capitalize()
-                contents.append(f'{key}: {val}')
+                contents.append(f"{key}: {val}")
+
+        return "\n".join(contents)
+
+    def orcid_metadata(self, obj):
+        """
+        Get metadata for orcid users.
+
+        This is the fast and dirty way, by checking substrings against the ORCiD
+        ID format. The actual ORCiD ID has the last digit as a checksum, and
+        strongly recommends using an OAUTH flow to get authenticated results.
+
+        The regex below CANNOT validate that an orcid ID exists, or points to the
+        correct user.
+        """
+        orcid_id_regex = r"\d{4}-\d{4}-\d{4}-(?:\d{3}X|\d{4})"
+
+        def get_orcid_ids(input_str, regex):
+            """Return all substrings matching regex."""
+            match = re.findall(regex, input_str)
+            logger.debug(f"ORCiD Match on {input_str} results in: {match}")
+            return match
+
+        def is_orcid_id(test_id, regex):
+            """Check if single ID is orcid ID."""
+            match = re.fullmatch(regex, test_id)
+            logger.debug(f"ORCiD Check on {test_id} results in: {match}")
+            return match
+
+        def get_orcid_url(orcid_id):
+            """Return url of form https://orcid.org/0000-0012-2345-6789."""
+            base_url = "https://orcid.org/"
+            if is_orcid_id(orcid_id, orcid_id_regex):
+                result = urllib.parse.urljoin(base_url, orcid_id)
+                logger.debug(f"converting {orcid_id} to {result}")
+                return result
+            else:
+                logger.debug(f"could not map {orcid_id} to URL")
+                return None
+
+        # Get all matching substrings
+        checked_ids = []
+        for user_id in [obj.requestor.username, obj.requestor.email]:
+            orcid_ids = get_orcid_ids(user_id, orcid_id_regex)
```

Comment:
This could be simplified using a few things in Python, namely itertools.chain and list comprehension.

```python
from itertools import chain
search_keys = [obj.requestor.username, obj.requestor.email]
orcid_ids = [oid for oid in chain(*[get_orcid_ids(search, orcid_id_regex) for search in search_keys) if oid]
```

The `if oid` is only necessary if the list can have None or falsy values in it.

Reply:
used flatten() to avoid needing to handle [ [], [], ] that regex would spit out occassionally.